esp/GBPlay/main/socket.c

```diff
@@ -0,0 +1,253 @@
+                break;
+            }
+
+            ESP_LOGE(__func__, "Socket unable to connect: errno %d", sock_error);
+            success = false;
+        }
+    }
+
+    if (errno == ETIMEDOUT)
+    {
+        ESP_LOGE(
+            __func__,
+            "Timed out waiting for socket connection after %d ms",
+            timeout_ms
+        );
+    }
+
+    return success;
+}
+
+static bool _connect_socket(int sock, struct addrinfo* address, int timeout_ms)
+{
+    // Temporarily switch to non-blocking so we can control the timeout
+    if (!_set_socket_is_blocking(sock, false))
+    {
+        ESP_LOGE(__func__, "Unable to set socket to non-blocking");
+        return false;
+    }
+
+    bool success = true;
+    if (connect(sock, address->ai_addr, address->ai_addrlen) < 0)
+    {
+        if (errno != EINPROGRESS)
+        {
+            ESP_LOGE(__func__, "Socket unable to connect: errno %d", errno);
+            success = false;
+        }
+        else
+        {
+            success = _wait_for_socket_connect(sock, timeout_ms);
+        }
+    }
+
+    if (!_set_socket_is_blocking(sock, true))
+    {
+        ESP_LOGE(__func__, "Unable to set socket to blocking");
+        success = false;
+    }
+
+    return success;
+}
+
+static int _create_socket(struct addrinfo* address, int timeout_ms)
+{
+    assert(address->ai_protocol == IPPROTO_TCP);
+
+    int sock = socket(address->ai_family, address->ai_socktype, address->ai_protocol);
+    if (sock < 0)
+    {
+        ESP_LOGE(__func__, "Unable to create socket: errno %d", errno);
+        return -1;
+    }
+
+    // Reduce latency
+    int nodelay_value = 1;
+    if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &nodelay_value, sizeof(nodelay_value)) != 0)
+    {
+        ESP_LOGE(__func__, "Unable to set socket options: errno %d", errno);
+        close(sock);
+        return -1;
+    }
+
+    if (!_connect_socket(sock, address, timeout_ms))
+    {
+        close(sock);
+        return -1;
+    }
+
+    return sock;
+}
+
+int socket_connect(const char* address, uint16_t port, int timeout_ms)
+{
+    struct addrinfo hints = {0};
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_protocol = IPPROTO_TCP;
+
+    char service[NI_MAXSERV] = {0};
+    snprintf(service, sizeof(service), "%d", port);
+    service[NI_MAXSERV - 1] = '\0';
+
+    struct addrinfo* address_info = NULL;
+    if (getaddrinfo(address, service, &hints, &address_info) != 0)
+    {
+        ESP_LOGE(__func__, "Could not get address info for %s:%d", address, port);
+        return -1;
+    }
+    else
+    {
+        int sock = -1;
+        for (struct addrinfo* a = address_info; a != NULL; a = a->ai_next)
+        {
+            sock = _create_socket(a, timeout_ms);
+            if (sock >= 0)
+            {
+                break;
+            }
+        }
+
+        freeaddrinfo(address_info);
+        return sock;
+    }
+}
+
+// TODO: detect unclean disconnect (read timeout)
```

Comment:
If we "pull the plug" on the server, the device can't detect the disconnect. This problem exists the other way around too. We can fix it with timeouts and ping messages.